### PR TITLE
Code Review v1.0

### PR DESCRIPTION
**Pull request by calebfavor**
I am attempting to do this through the github code review functionality. In order to do so, I created a pull request for the entire master branch onto a new empty branch. This should let me review the entire repo. We'll see :)

### CODE REVIEW

File: src/Controllers/AuthController.php [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+use Railroad\Usora\Entities\UserFirebaseTokens;
+use Railroad\Usora\Managers\UsoraEntityManager;
+use Railroad\Usora\Repositories\UserFirebaseTokensRepository;
+use Railroad\Usora\Repositories\UserRepository;
+use Tymon\JWTAuth\JWTAuth;
+
+class AuthController extends Controller
+{
+    /**
+     * @var EntityManager
```
Comment (calebfavor) [inline src/Controllers/AuthController.php]: There are some missing php use statements and some extra ones not being used. This likely needs some cleanup across the entire project regarding use statements.

File: src/Controllers/ContentController.php [hunk 2]
```diff
@@ -0,0 +1,651 @@
...
+
+        ModeDecoratorBase::$decorationMode = DecoratorInterface::DECORATION_MODE_MINIMUM;
+
+        $parentChildrenTrimmed = $this->getParentChildTrimmed($parentChildren, $content);
+
+        $content['related_lessons'] = $parentChildrenTrimmed;
+
+        $content = $this->attachNextPrevLesson($parent, $content, $parentChildren);
+
+        CommentRepository::$availableContentId = $content['id'];
```
Comment (calebfavor) [inline src/Controllers/ContentController.php:179]: Some of these controller methods could use some better spacing and some comments to explain what is going on. I had a hard time following these functions.

File: src/Controllers/UserProgressController.php [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+            $parent = $this->contentService->getByChildId($request->get('content_id'));
+        }
+
+        $response =  [
+            'success' => true,
+            'parent' => $parent,
+        ];
+
+        if(config('musora-api.shouldDisplayReview', false)) {
+            $user = $this->userRepository->findOneBy(['id' => auth()->id()]);
```
Comment (calebfavor) [inline src/Controllers/UserProgressController.php]: Some of these class attributes do not exist. I am not sure this controller will function.

File: src/Transformers/ContentTransformer.php [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+            $responseStructure = config('musora-api.response-structure.' . $type);
+
+        }
+
+        if (!$responseStructure) {
+            return (!is_array($content)) ? $content->getArrayCopy() : $content;
+        }
+
+        foreach ($responseStructure as $index => $value) {
```
Comment (calebfavor) [inline src/Transformers/ContentTransformer.php:50]: There may be an opportunity to refactor some of this type of logic among all the transformers. I also found it very hard to follow what's going on. Some explanation comments would be useful

File: src/Controllers/AuthController.php [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+            $this->userProvider->setCurrentUserProfilePictureUrl($request->get('file'));
+        }
+
+        if ($request->has('phone_number')) {
+            $this->userProvider->setCurrentUserPhoneNumber($request->get('phone_number'));
+        }
+
+        if ($request->has('display_name')) {
+            $updatedUser = $this->userProvider->setCurrentUserDisplayName($request->get('display_name'));
+            if (!$updatedUser) {
```
Comment (ghost) [inline src/Controllers/AuthController.php]: Controller methods multiple return statements, and a few if statements, may be replaced with custom exceptions. This would increase code readability and decreases complexity.
Example of custom exception: https://github.com/railroadmedia/railcontent/blob/1.0-/src/Exceptions/NotAllowedException.php

File: src/Controllers/AuthController.php [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        try {
+            event(
+                new AppSignupStartedEvent(
+                    [
+                        'email' => $request->get('email'),
+                    ]
+                )
+            );
+        } catch (Exception $e) {
+            return response()->json(
```
Comment (ghost) [inline src/Controllers/AuthController.php]: Instead of having a 2nd controller method return statement, a custom exception may be thrown here or in AppSignupStartedEvent event handlers.

File: src/Controllers/AuthController.php [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+use Exception;
+use Illuminate\Http\JsonResponse;
+use Illuminate\Http\Request;
+use Illuminate\Routing\Controller;
+use Railroad\Ecommerce\Events\AppSignupStartedEvent;
+use Railroad\MusoraApi\Contracts\UserProviderInterface;
+use Railroad\MusoraApi\Requests\CreateIntercomUserRequest;
+
+class AuthController extends Controller
```
Comment (ghost) [inline src/Controllers/AuthController.php:17]: Controller class name may be updated to reflect all controller actions. AvatarController action may be refactored in here, to keep all user account related actions in one place.

File: src/Controllers/ContentController.php [hunk 8]
```diff
@@ -0,0 +1,638 @@
...
+     * @return JsonResponse
+     * @throws Exception
+     */
+    public function submitQuestion(SubmitQuestionRequest $request)
+    {
+        $input = $request->all();
+        $currentUser = $this->userProvider->getCurrentUser();
+        $brand = config('railcontent.brand', '');
+
+        $input['subject'] =
```
Comment (ghost) [inline src/Controllers/ContentController.php:428]: Extensive usage of array maps instead of objects. Unlike PHP 4.x/5.x, objects usage does not have a negative impact on PHP 7.x performance. I did looked into mailora package and sadly I did not found any model/entity/interface.

File: src/Controllers/ContentController.php [hunk 9]
```diff
@@ -0,0 +1,638 @@
...
+     * @param Request $request
+     * @return JsonResponse
+     * @throws NonUniqueResultException
+     */
+    public function getContent($contentId, Request $request)
+    {
+        $content = $this->contentService->getById($contentId);
+
+        if (!$content) {
+            return response()->json();
```
Comment (ghost) [inline src/Controllers/ContentController.php]: I think a proper custom exception may be thrown, instead of having second empty response return statement.

File: src/Controllers/PacksController.php [hunk 10]
```diff
@@ -0,0 +1,536 @@
...
+    public function jumpToNextLesson(Request $request, $id)
+    {
+        $pack = $this->contentService->getById($id);
+
+        $thisPackBundle = null;
+
+        if ($pack['type'] == 'pack') {
+
+            $lessonsProgressRows =
+                $this->contentRepository->connection()
```
Comment (ghost) [inline src/Controllers/PacksController.php]: I think this query may be refactored into a repository or service.

File: src/Controllers/UserProgressController.php [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+            'device_type' => 'required',
+        ];
+
+        $validator = Validator::make(
+            $request->all(),
+            $rules
+        );
+
+        if ($validator->fails()) {
+            return response()->json(
```
Comment (ghost) [inline src/Controllers/UserProgressController.php]: I think this validation logic may be refactored into request validation class.

File: src/Controllers/LearningPathController.php [hunk 12]
```diff
@@ -0,0 +1,487 @@
...
+    public function showLesson($lessonId)
+    {
+        ContentRepository::$bypassPermissions = true;
+        ContentRepository::$availableContentStatues = false;
+        ContentRepository::$pullFutureContent = true;
+
+        $thisLesson = $this->contentService->getById($lessonId);
+
+        if (empty($thisLesson)) {
+            return response()->json($thisLesson);
```
Comment (ghost) [inline src/Controllers/LearningPathController.php]: I think some of the 5 method return statements may be replaced with not found exception.

File: tests/Content/ContentControllerTest.php [hunk 13]
```diff
@@ -0,0 +1,558 @@
...
+
+    public function test_pull_not_existing_content_endpoint()
+    {
+        $response =
+            $this->actingAs()
+                ->get(
+                    '/musora-api/content/' . rand()
+                );
+
+        $response->assertStatus(200);
```
Comment (ghost) [inline tests/Content/ContentControllerTest.php]: I think this should be a 404. If a related content does not exist, the content that linked to it should find its way to an error log to be reviewed and updated.

File: src/Controllers/PacksController.php (hunk 10 above)
Comment (roxanariza) [reply src/Controllers/PacksController.php]: Agree. Moved the query into a service. Thanks!

**Review by JonathanTM — APPROVED** [review]
